cadc-dali/src/main/java/ca/nrc/cadc/dali/tables/parquet/DynamicSchemaGenerator.java

```diff
@@ -0,0 +1,72 @@
+package ca.nrc.cadc.dali.tables.parquet;
+
+import ca.nrc.cadc.dali.tables.votable.VOTableField;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.apache.avro.Schema;
+import org.apache.log4j.Logger;
+
+public class DynamicSchemaGenerator {
+
+    private static final Logger log = Logger.getLogger(DynamicSchemaGenerator.class);
+
+    public static Schema generateSchema(List<VOTableField> voFields) {
+        // List to hold Avro fields
+        List<Schema.Field> fields = new ArrayList<>();
+        try {
+            int columnCount = voFields.size();
+            log.debug("VOTable Column count = " + columnCount);
+            for (VOTableField voField : voFields) {
+                String columnName = voField.getName();
+                Schema.Field field = new Schema.Field(columnName.replaceAll("\"", ""), getAvroFieldType(voField.getDatatype()), null, null);
+                fields.add(field);
+            }
+            log.debug("Avro Schema.Field count = " + fields.size());
+        } catch (Exception e) {
+            log.debug("Failure while creating Avro Schema from VOTable", e);
+            throw new RuntimeException("Failure while creating Avro Schema from VOTable : " + e.getMessage(), e);
+        }
+
+        // Define the Avro record schema with the fields
+        Schema schema = Schema.createRecord("Record", null, null, Boolean.FALSE);
+        schema.setFields(fields);
+        log.debug("Schema Generated Successfully : " + schema);
+        return schema;
+    }
+
+    private static Schema getAvroFieldType(String sqlType) {
```

Comment:
First time looking at this file... everything else looks OK.

I think this method needs to look at more of the VOTableField metadata. The complete datatype is defined by `datatype`, `arraysize`, and `xtype`. The latter two are optional and if null it means the data is a single primitive.

For "char" that actually means a single char; a string is technically `datatype="char" arraysize="X"` (for some value of X allowed by the VOTable spec). Maybe those are the same to avro. 

A few xtypes like "uri" and "uuid" you could probably ignore for now, but we should try to do "timestamp" correctly. We can also ignore geometry for now (point, circle, polygon, etc) but will need to address if before release.

So the changes: change method signature to `getAvroFieldType(VOTableField)`, check the char vs string correctness, and fix the timestamp (it's a char array with an xtype):

`datatype="char" arraysize="X" xtype="timestamp"`

In that column, the value objects should be java.util.Date and I don't know if avro will format that correctly in a standard way or if we need to format it in a VO-compatible way. Any idea?

Reply:
iirc, in miost of our TAP services timestamp columns have this metadata:
```
datatype="char" arraysize="23*" xtype="timestamp"
```
that arraysize means variable length with a max of 23 characters... just noting that you probably do not need to look at the value of arraysize (at least for this one), just the xtype.

---

In future, if you need to know the arraysize it "is complicated" because length and variable-size are both embedded in there, and in a few cases there are even multi-dimensional arrays. An example of that is the caom2.Plane.energy_bounds_samples column. The datatype will be Interval[] or List<Interval> and in VOTable that looks like
```
datatype="double" arraysize="2x*" xtype="interval"
```
where 2x* means a 2 x var-length array aka an array of 2-value intervals.
We only have a few cases of multidimensional: